vscript vgui

samisalreadytaken opened the pull request:
This implementation is not final, but it is stable.

Example code and videos:

https://github.com/samisalreadytaken/vscripts/tree/master/mapbase
https://gist.github.com/samisalreadytaken/4847ede81a7222b66dc82eac8622904e

https://youtu.be/PrND1JsXduo
https://youtu.be/I4Y_kLR2CQ0
https://youtu.be/Oh5JpBHXWYM

Some notes:

- Most `#if 0`'d blocks are either untested, or usefulness of them are uncertain.

- Should the function to set overrides/callbacks be named `SetCallback()`, or something else? Panorama uses `SetPanelEvent()` for registering callbacks.

- A future change to add is integration with `C_VGuiScreen` entity to display script panels. This would also be a simpler alternative to manual 3D world painting.

- A TODO is a convenient method to control the visibility of existing HUD elements. The current hack is to use the `hidehud` convar which is manually reset on level shutdown. (See [hud_override.nut](https://github.com/samisalreadytaken/vscripts/blob/master/mapbase/hud_override.nut))

---

#### PR Checklist
- [x] **My PR follows all guidelines in the CONTRIBUTING.md file**
- [x] My PR targets a `develop` branch OR targets another branch with a specific goal in mind

z33ky reviewed (commented):
There are minor fomatting issues with inconsistent spacing around parenthesis in control-flow statements and function parameters. It's not really concerning, just pointing it out.
Missing spaces after `SCRIPT_SINGLETON` (`BEGIN_SCRIPTDESC_ROOT_NAMED()`) also caught my eye.

The name of `_panel` diverges from the more common style (`m_pPanel`), though it has precedent in the original VGUI code.
This and `m_hScriptInstance` have accessors and could become `private`; I think only `ResolveChildren()` would need to be modified, either by being `friend`ed or by becoming a static method of `IScriptVGUIObject`. Again, just pointing it out.

There are a couple of loops to search for matching panels in `g_ScriptPanels`. That could be refactored to a function.

There are two unrelated changes in `sp/src/game/client/vscript_client.cpp` which should get their own dedicated commit (replacement of loop with `V_memset()` and removal of a call to `g_pScriptVM->ClearValue( m_ScriptScope, "entity" )` in `CScriptMaterialProxy::OnBind()`).

There are a couple of places where references to a `static Vector` is returned.
I'm not sure how the script FFI deals with these - if a script were to call it twice, would both variables in the script also refer to the same object? I.e. would something like this fail:
```
local top = ScreenToWorld(0.5, 0);
local bottom = ScreenToWorld(0.5, 1);
assert( top != bottom );
```

samisalreadytaken commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1541 @@
+
+static inline int HandleToInt( unsigned int i )
+{
+	Assert( i <= INT_MAX );
```
These handles are vector indices, and are in tens or hundreds. I'm not sure what would happen if it were to exceed INT_MAX in a malicious attempt because CUtlVector works with signed int.

This check here is not necessary because internally the returned values are signed. The assert is just redundancy.

samisalreadytaken commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1541 @@
+	{
+		int x, y;
+		ipanel()->GetPos( GetVPanel(), x, y );
+		g_pScriptVM->SetValue( out, (ScriptVariant_t)0, x );
```
Ambigious call with `SetValue(HSCRIPT, const char*, ScriptVariant*)` results in VM stack corruption.

z33ky replied:
I see. Doesn't even produce a warning, although it looks like a nice footgun.

samisalreadytaken commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1541 @@
+class IScriptVGUIPanel
+{
+public:
+	void Shutdown() {}
```
`Shutdown()` _can_ be manually defined in all panel classes, but there are more than one that is going to be empty. It is there to reduce repetition, to implement only when needed; and also a remnant of an older iteration of the code that could access script panel info from `IScriptVGUIObject`.

Do you think putting empty functions in each class is better?

z33ky replied:
> Do you think putting empty functions in each class is better?

Adding `: public IScriptVGUIPanel` is about the same amount of boilerplate as writing `void Shutdown() {}`, though I don't have a strong preference. I think a short comment why this isn't `virtual` even though it's an "Interface"-class would suffice, or perhaps also just renaming `IScriptVGUIPanel` to something like `CScriptVGUIPanelStub` or `CEmptyScriptVGUIPanel`.

samisalreadytaken commented:
`_vpanel` and `m_hScriptInstance` are also accessed from derived helper classes and from `CScriptVGUI::Create()`. There _could_ be setter functions, but I don't think it really matters.

Returned Vectors to script functions are reallocated in the VM, returning static Vector is Valve's convention. Though testing local Vector return, it seems to work fine; perhaps it's a difference in the implementation, or I'm forgetting something.

I will push some changes later.

z33ky commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1871 @@
+
+static const char *GetFixedFontName( const char *name, bool proportional )
+{
+	static char fontName[64];
```
It'd be cleaner to take `char *fontName, size_t fontNameSize` parameters.

z33ky commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1871 @@
+
+	void OnTick()
+	{
+		if ( m_nLastFrame >= gpGlobals->framecount )
```
Why is this check necessary?

samisalreadytaken replied:
`OnTick()` is called twice in the same frame before the next.

z33ky commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1871 @@
+
+void CScriptSurface::DrawOutlinedRect( int x0, int y0, int width, int height, int thickness )
+{
+	//surface()->DrawOutlinedRect( x0, y0, x0 + width, y0 + height );
```
Leftover

z33ky commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1871 @@
+	}
+	else
+	{
+		int size = V_strlen( windowsFontName ) + 1;
+		char *fontName = new char[ size ];
+		V_strncpy( fontName, windowsFontName, size );
```
```suggestion
		char *fontName = strdup( windowsFontName );
```
Free via `free()` instead `delete[]`.

Some other places truncate the fontname to 64 bytes, maybe `FontData_t::name` could also become `char[64]` so no memory management is necessary.

samisalreadytaken replied:
That 64 byte long string is the font alias allocated in the dictionary, not the font name.

z33ky commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1871 @@
+		scanlines = scheme()->GetProportionalScaledValue( scanlines );
+	}
+
+#ifdef _DEBUG
+	bool bSuccess =
+#endif
+
+		surface()->SetFontGlyphSet( font, windowsFontName, tall, weight, blur, scanlines, flags );
+
+#ifdef _DEBUG
+	if ( bSuccess )
+	{
+		Msg( "Create font [%i]%s\n", font, fontAlias );
+	}
+	else
+	{
+		Warning( "Failed to create font [%i]%s\n", font, fontAlias );
+	}
+#endif
```
Indentation looks weird. May I suggest
```suggestion
	bool bSuccess = surface()->SetFontGlyphSet( font, windowsFontName, tall, weight, blur, scanlines, flags );

#ifdef _DEBUG
	if ( bSuccess )
	{
		Msg( "Create font [%i]%s\n", font, fontAlias );
	}
	else
	{
		Warning( "Failed to create font [%i]%s\n", font, fontAlias );
	}
#else
	(void)bSuccess;
#endif
```

z33ky commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1871 @@
+	VPANEL _vpanel;
+	HSCRIPT m_hScriptInstance;
+
+	// Called in deletion
```
```suggestion
	// Called on deletion (CScriptVGUIObject::Destroy())
```

z33ky commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1871 @@
+{
+	FOR_EACH_VEC( g_ScriptPanels, i )
+	{
+		g_ScriptPanels[i]->Destroy();
```
Maybe `IScriptVGUIObject::~IScriptVGUIObject()` should call `Destroy()`.

samisalreadytaken replied:
It causes a build error. I didn't think it was important enough to look into.

z33ky replied:
Ah, calling virtual functions in the destructor (also constructor) works like static dispatch, so the derived methods wouldn't be invoked.

samisalreadytaken commented on sp/src/game/client/mapbase/vscript_vgui.cpp:
```diff
@@ -0,0 +1,1871 @@
+	return static_cast< unsigned int >(i);
+}
+
+static inline int HandleToInt( unsigned int i )
```
This can be used for other vgui handles such as HScheme, HContext and HTexture. I think it's fine for it to stay generic.

z33ky replied:
That's fair,

Blixibon reviewed (changes requested):
There's not any practical way for me to test everything this has, although I did try using the HUD replacement scripts from your repo and messed around with the save icon script you provided. I experienced an `Overflowed reading usercmd data (check sending and receiving code for mismatches)!` error from the fake Steam client, which I'm assuming isn't related to the VScript VGUI code itself, but I'm otherwise in agreement that this seems stable. Before I approve this, I'm requesting a change I had to make for it to compile on my machine.

I'm really blown away by the capabilities of this implementation. This has the potential to open custom VGUI creation to a much wider audience.

---

> Should the function to set overrides/callbacks be named `SetCallback()`, or something else? Panorama uses `SetPanelEvent()` for registering callbacks.

The context of `SetCallback` is specific enough that I think the name works best as-is than it would as something more explicit.

samisalreadytaken commented:
The Steam notifications & achievements code works fine, are you sure you've included it on both server and client, and there are no errors on server? That error happens when client sends a message and server doesn't receive it. I also updated that code for save/restore few days ago (and rebased because they're code for an unreleased WIP feature so I don't care about history), although unlikely, you might have used a broken version.

Additionally I removed `WorldToScreen()` for now because I'm not sure about its output parameter. I had initially chosen 0,1 indices for an array because arrays are the lightest containers in Squirrel, but it could use x,y as well. The HSCRIPT parameter memory leak is also unideal, people shouldn't really use it.

Not too worried about temporarily removing this because it's [already possible](https://github.com/samisalreadytaken/vs_library/blob/master/Documentation.md#f_WorldToScreen) to manually calculate this.

Blixibon commented:
Ah, I didn't realize it was meant to run on both the server and the client. I saw that the save icon script was client only and I didn't realize your other HUD scripts were meant to be shared. I'm surprised I missed that, but sorry for the confusion.

samisalreadytaken commented:
The Steam scripts are only shared for the achievement manager (serverside control and verification), and the HUD scripts are shared to be able to get suit information because NetProps doesn't work.
